server-common/src/main/java/org/apache/kafka/admin/AdminUtils.java

```diff
@@ -0,0 +1,257 @@
+     * each rack will get at least one replica. Otherwise, each rack will get at most one replica. In a perfect
+     * situation where the number of replicas is the same as the number of racks and each rack has the same number of
+     * brokers, it guarantees that the replica distribution is even across brokers and racks.
+     * </p>
+     * @return a Map from partition id to replica ids
+     * @throws AdminOperationException If rack information is supplied but it is incomplete, or if it is not possible to
+     *                                 assign each replica to a unique rack.
+     *
+     */
+    public static Map<Integer, List<Integer>> assignReplicasToBrokers(Collection<BrokerMetadata> brokerMetadatas,
+                                                                      int nPartitions,
+                                                                      int replicationFactor,
+                                                                      int fixedStartIndex,
+                                                                      int startPartitionId) {
+        if (nPartitions <= 0)
+            throw new InvalidPartitionsException("Number of partitions must be larger than 0.");
+        if (replicationFactor <= 0)
+            throw new InvalidReplicationFactorException("Replication factor must be larger than 0.");
+        if (replicationFactor > brokerMetadatas.size())
+            throw new InvalidReplicationFactorException("Replication factor: " + replicationFactor + " larger than available brokers: " + brokerMetadatas.size() + ".");
+        if (brokerMetadatas.stream().noneMatch(b -> b.rack.isPresent()))
+            return assignReplicasToBrokersRackUnaware(nPartitions, replicationFactor, brokerMetadatas.stream().map(b -> b.id).collect(Collectors.toList()), fixedStartIndex,
+                startPartitionId);
+        else {
+            if (brokerMetadatas.stream().anyMatch(b -> !b.rack.isPresent()))
+                throw new AdminOperationException("Not all brokers have rack information for replica rack aware assignment.");
+            return assignReplicasToBrokersRackAware(nPartitions, replicationFactor, brokerMetadatas, fixedStartIndex,
+                startPartitionId);
+        }
+    }
+
+    private static Map<Integer, List<Integer>> assignReplicasToBrokersRackUnaware(int nPartitions,
+                                                                                  int replicationFactor,
+                                                                                  Collection<Integer> brokerList,
+                                                                                  int fixedStartIndex,
+                                                                                  int startPartitionId) {
+        Map<Integer, List<Integer>> ret = new HashMap<>();
+        Integer[] brokerArray = brokerList.toArray(new Integer[0]);
```

Comment:
Do we need to create this intermediate array? Can't we use `brokerList`, it seems the callers pass in `List`